base/binaryplatforms.jl

```diff
@@ -0,0 +1,999 @@
+       select_platform, platforms_match, platform_name
+import .Libc.Libdl
+
+# This exists to ease compatibility with old-style Platform objects
+abstract type AbstractPlatform; end
+
+"""
+    Platform
+
+A `Platform` represents all relevant pieces of information that a julia process may need
+to know about its execution environment, such as the processor architecture, operating
+system, libc implementation, etc...  It is, at its heart, a key-value mapping of tags
+(such as `arch`, `os`, `libc`, etc...) to values (such as `"arch" => "x86_64"`, or
+`"os" => "windows"`, etc...).  `Platform` objects are extensible in that the tag mapping
+is open for users to add their own mappings to, as long as the mappings do not conflict
+with the set of reserved tags: `arch`, `os`, `os_version`, `libc`, `call_abi`,
+`libgfortran_version`, `libstdcxx_version`, `cxxstring_abi` and `julia_version`.
+
+Valid tags and values are composed of alphanumeric and period characters.  All tags and
+values will be lowercased when stored to reduce variation.
+
+Example:
+
+    Platform("x86_64", "windows"; cuda = "10.1")
+"""
+struct Platform <: AbstractPlatform
+    tags::Dict{String,String}
+    # The "compare strategy" allows selective overriding on how a tag is compared
+    compare_strategies::Dict{String,Any}
+
+    function Platform(arch::String, os::String;
+                      validate_strict::Bool = false,
+                      compare_strategies::Dict{String,Any} = Dict{String,Any}(),
+                      kwargs...)
+        # A wee bit of normalization
+        os = lowercase(os)
+        arch = lowercase(arch)
+        if arch ∈ ("amd64",)
+            arch = "x86_64"
+        elseif arch ∈ ("i386", "i586")
+            arch = "i686"
+        elseif arch ∈ ("arm",)
+            arch = "armv7l"
+        elseif arch ∈ ("ppc64le",)
+            arch = "powerpc64le"
+        end
+
+        tags = Dict{String,String}(
+            "arch" => arch,
+            "os" => os,
+        )
+        for (tag, value) in kwargs
+            tag = lowercase(string(tag))
+            if tag ∈ ("arch", "os")
+                throw(ArgumentError("Cannot double-pass key $(tag)"))
+            end
+
+            # Drop `nothing` values; this means feature is not present or use default value.
+            if value === nothing
+                continue
+            end
+
+            # Normalize things that are known to be version numbers so that comparisons are easy.
+            # Note that in our effort to be extremely compatible, we actually allow something that
+            # doesn't parse nicely into a VersionNumber to persist, but if `validate_strict` is
+            # set to `true`, it will cause an error later on.
+            if tag ∈ ("libgfortran_version", "libstdcxx_version", "os_version")
+                normver(x::VersionNumber) = string(x)
+                function normver(str::AbstractString)
```

Comment:
Perhaps make this

```suggestion
                function normver(str::String)
```

and then call it with `normver(string(value)::String)` below?